calblueprint/ogsc
# Validations: field key-specific constraints, inline error messaging, and more helpful error labeling

This PR addresses a number of issues with validations as they were handled in profile field create/update ops and in the create player form. 

* Adds field key-specific constraints (such as min/max for Engagement Scores, GPA). Closes #222, closes #228, closes  closes #229, closes #230, closes #232.
* Shows inline error messaging with form inputs in Create Player form and in ProfileSection (updating existing profile). Closes #224, closes #227.
* Makes the labeling of validation errors from the API better across the board (using `labelProfileField` instead of "value" or "profileFields[6].value").

## TODO

- [x] ProfileFieldEditorModal not showing errors inline with the value input (it appears at the bottom of the modal instead)
- [x] "Save" actions are not being prevented in ProfileFieldEditorModal when validation errors appear
- [x] UX for understanding validation errors and addressing them later could be improved in the Create Player form
  - "Preventing from moving onto the next step" of the form has been suggested, but I think this might be non-ideal as there are many ways to jump between steps of the form (progress bar at top, back/forward buttons at bottom, browser navigation stack) - some of which we can't actually control for (like URL changes).
  - Instead, I think we should consider showing the section in an invalid state within the progress bar, with a warning icon/popover combo that suggests that there are "X number of errors" to address in that specific section.

## How to Test/Use PR feature

* Regression tests for #222, #228,  #229, #230, #232. Attempt to create invalid inputs in the Create Player form or in the player profile (to update fields).
* Stress test invalid inputs to see if there are other field key-specific constraints we want to add across the app.
* Try to ignore inline validation errors and submit the player form at the very end. Observe the change in error messaging.

## PR Checklist

Does your branch (check if true):

## Screenshots

Inline validation error, in the create form
<img src="https://user-images.githubusercontent.com/2977329/116032236-2679b800-a614-11eb-80e8-f447331a5f53.png" height="150">

Inline validation errors, in the player profile
<img src="https://user-images.githubusercontent.com/2977329/116032346-5b860a80-a614-11eb-83af-c686b7188307.png" height="300">


CC: @sydbui
